api/lib/domain/models/CampaignTubeRecommendation.js

```diff
@@ -1,26 +1,53 @@
+const recommendationService = require('../services/recommendation-service');
+
+const _ = require('lodash');
+const fp = require('lodash/fp');
+
 class CampaignTubeRecommendation {
 
   constructor({
     // attributes
     campaignId,
-    tubeId,
-    competenceId,
-    competenceName,
-    tubePracticalTitle,
-    areaColor,
+    tube,
+    competence,
+    skills,
+    validatedKnowledgeElements,
+    numberOfParticipants,
+    maxSkillLevelInTargetProfile
   } = {}) {
     // attributes
     this.campaignId = campaignId;
-    this.tubeId = tubeId;
-    this.competenceId = competenceId;
-    this.competenceName = competenceName;
-    this.tubePracticalTitle = tubePracticalTitle;
-    this.areaColor = areaColor;
+    this.tubeId = tube.id;
+    this.competenceId = competence.id;
+    this.competenceName = competence.name;
+    this.tubePracticalTitle = tube.practicalTitle;
+    this.areaColor = competence.area.color;
+    this.averageScore = null;
+    if (numberOfParticipants) {
+      this.averageScore = this._computeAverageScore(validatedKnowledgeElements, skills, numberOfParticipants, maxSkillLevelInTargetProfile);
+    }
   }
 
   get id() {
     return `${this.campaignId}_${this.tubeId}`;
   }
+
+  _computeAverageScore(validatedKnowledgeElements, skills, numberOfParticipants, maxSkillLevelInTargetProfile) {
+    return fp.flow(
+      _.curry(this._knowledgeElementsByParticipant)(validatedKnowledgeElements),
+      fp.map(_.curry(recommendationService.computeRecommendationScore)(skills, maxSkillLevelInTargetProfile)),
+      fp.reduce((total, score) => total + score, 0),
+      (sumOfScores) => sumOfScores / numberOfParticipants,
+    )(numberOfParticipants);
```

Comment:
je ne connaissais pas `lodash/fp`, qu'est-ce qui vous a motivé à l'utiliser ici ?

Reply:
C'est intéressant comme écriture. J'ai l'impression qu'en restant dans `lodash/fp` on peut un peu alléger (élimination de `_knowledgeElementsByParticipant`, `curry` -> `bindKey`, `reduce` -> `sum`, et la valeur qui rentre dans le tuyau n'est plus `numberOfParticipants` mais `validatedKnowledgeElements` ce qui me paraît faire plus de sens) : 

```js
_computeAverageScore(validatedKnowledgeElements, skills, numberOfParticipants, maxSkillLevelInTargetProfile) {
    return fp.flow(
      fp.groupBy('userId'),
      fp.values,
      fp.map(_.bindKey(recommendationService, 'computeRecommendationScore', skills, maxSkillLevelInTargetProfile)),
      fp.sum
    )(validatedKnowledgeElements) / numberOfParticipants;
}
```

Noter en particulier que l'utilisation de `_.curry` sur une méthode d'objet (ici `recommendationService.computeRecommendationScore`) n'est pas _safe_ parce que le `this` reçu par la méthode n'est pas le bon ! La fonction `_.bindKey` évite cela.

J'aime bien `lodash/fp` mais dans ce cas j'aurais quand même une préférence pour une version plus "basique" :

```js
_computeAverageScore(validatedKnowledgeElements, skills, numberOfParticipants, maxSkillLevelInTargetProfile) {
  const sumOfScores = _.groupBy(validatedKnowledgeElements, 'userId')
    .values()
    .sumBy((knowledgeElements) =>
	  recommendationService.computeRecommendationScore(
	    skills,
	    maxSkillLevelInTargetProfile,
	    knowledgeElements
	  ));

  return sumOfScores / numberOfParticipants;
}
```

Reply:
Ok on repasse sur une version basique

Reply:
@ajoanny 👍 

Dommage d'avoir laissé `_knowledgeElementsByParticipant` qui ne sert plus à rien (le `toArray()` est redondant avec le `values()` fait ensuite, l'ajout des _n_ listes vides est superflu) 😿 ?